slashing-spec.md

```diff
@@ -0,0 +1,46 @@
+This file names and documents the various slashing conditions we use in Gravity.
+
+## GRAVSLASH-01: Signing fake validator set or tx batch evidence
+
+This slashing condition is intended to stop validators from signing over a validator set and nonce that has never existed on Cosmos. It works via an evidence mechanism, where anyone can submit a message containing the signature of a validator over a fake validator set. This is intended to produce the effect that if a cartel of validators is formed with the intention of submitting a fake validator set, one defector can cause them all to be slashed.
+
+**Implementation considerations:**
+
+The trickiest part of this slashing condition is determining that a validator set has never existed on Cosmos. To save space, we will need to clean up old validator sets. We could keep a mapping of validator set hash to true in the KV store, and use that to check if a validator set has ever existed. This is more efficient than storing the whole validator set, but its growth is still unbounded. It might be possible to use other cryptographic methods to cut down on the size of this mapping. It might be OK to prune very old entries from this mapping, but any pruning reduces the deterrence of this slashing condition.
+
+## GRAVSLASH-02: Failure to sign validator set update or tx batch
+
+This slashing condition is triggered when a validator does not sign a validator set update or transaction batch which is produced by the Gravity Cosmos module. This prevents two bad scenarios- 
+
+1. A validator simply does not bother to keep the correct binaries running on their system,
+2. A cartel of >1/3 validators unbond and then refuse to sign updates, preventing any validator set updates from getting enough signatures to be submitted to the Gravity Ethereum contract. If they prevent validator set updates for longer than the Cosmos unbonding period, they can no longer be punished for submitting fake validator set updates and tx batches (GRAVSLASH-01 and GRAVSLASH-02). 
+
+To deal with scenario 2, GRAVSLASH-02 will also need to slash validators who are no longer validating, but are still in the unbonding period. This means that when a validator leaves the validator set, they will need to keep running their equipment for 2 weeks. This is unusual for a Cosmos chain, and may not be accepted by the validators. Research is ongoing for ways to allow validators to stop signing before the unbonding period is fully over.
```

Comment:
I think they only need to keep signing for 24 hours, but the full unbonding period is fine for this first draft.

I do think we're missing an important condition here, mainly what to do when a validator who was jailed for not singing validator sets unjails themselves. Are they supposed to run their orchestrator before unjailing themselves? thus catching up? Otherwise the risk is that they miss validator sets and unjail themselves just to be flashed again a few blocks later when another set comes due.

Reply:
Good question